edx_name_affirmation/signals.py

```diff
@@ -0,0 +1,179 @@
+            verified_name=full_name,
+            proctored_exam_attempt_id=None,
+            verification_attempt_id=None
+        ).order_by('-created').first()
+        if verified_name:
+            # if there is a verified name that matches the user and name,
+            # note the attempt id and update the status if needed
+            verified_name.verification_attempt_id = attempt_id
+            if translated_status:
+                verified_name.status = status
+            verified_name.save()
+            log_str = 'Updated VerifiedName for user={user_id} to have status={status} and ' \
+                      'verification_attempt_id={attempt_id} '
+        else:
+            # We shouldn't get to this case, but if there is no verified name associated with the attempt id or name,
+            # create a new verified name
+            user = User.objects.get(id=user_id)
+            verified_name = VerifiedName.objects.create(
+                user=user,
+                verified_name=full_name,
+                profile_name=profile_name,
+                verification_attempt_id=attempt_id,
+                status=(translated_status if translated_status else VerifiedNameStatus.PENDING),
+            )
+            log_str = 'Created VerifiedName for user={user_id} to have status={status} ' \
+                      'and verification_attempt_id={attempt_id}'
+    if log_str:
+        log.info(
+            log_str.format(
+                user_id=user_id,
+                attempt_id=attempt_id,
+                status=verified_name.status
+            )
+        )
+
+
+def update_verified_name_for_proctoring_attempt(
+    attempt_id,
+    user_id,
+    status,
+    full_name,
+    profile_name,
+    is_onboarding,
+    is_proctored,
+    backend_supports_onboarding,
+    **kwargs
+):
+    """
+    Receiver for proctored exam attempt updates.
+
+    Args:
+        attempt_id(int): ID associated with the proctored exam attempt
+        user_id(int): ID associated with the proctored exam attempt's user
+        status(str): status in proctoring language for the proctored exam attempt
+        full_name(str): name to be used as verified name
+        profile_name(str): user's current profile name
+        is_onboarding(boolean): if the exam attempt is for an onboarding exam
+        is_proctored(boolean): if the exam attempt is for a proctored exam
+        backend_supports_onboarding(boolean): if the exam attempt is for an exam with a backend that supports onboarding
+    """
+    # We only care about updates from onboarding exams, or from proctored exams with a backend that does  not
+    # support onboarding. This is because those two event types are guaranteed to contain verification events,
+    # whereas timed exams and proctored exams with a backend that does support onboarding are not guaranteed
+    # to contain verification events.
+    if is_onboarding or (is_proctored and not backend_supports_onboarding):
```

Comment:
this function is nested enough to be really challenging

I'd suggest first turning this outer if into an early return